Edits for CDN 2021 chapter (#2584)
HTTPArchive/almanac.httparchive.org: src/content/en/2021/cdn.md

These are the limits to our testing [methodology](./methodology):

- **Simulated network latency:** We use a dedicated network connection that synthetically shapes traffic.

- **Single geographic location:** Tests are run from a single datacenter and cannot test the geographic distribution of many CDN vendors.

- **Cache effectiveness:** Each CDN uses proprietary technology and many, for security reasons, do not expose cache performance.

- **Localization and internationalization:** Just like geographic distribution, the effects of language and geo-specific domains are also opaque to these tests.

- **CDN detection:** This is primarily done through DNS resolution and HTTP headers. Most CDNs use a DNS CNAME to map a user to an optimal datacenter. However, some CDNs use Anycast IPs or direct A+AAAA responses from a delegated domain which hide the DNS chain. In other cases, websites use multiple CDNs to balance between vendors, which is hidden from the single-request pass of our crawler.

> All of this influences our measurements.

Reviewer: I toned this down a litte. Hope you don't mind @Navaneeth-akam

Reviewer 2: Yup....looks good.

Most importantly, these results reflect the utilization of specific features (Example: TLS, HTTP/2 etc.,) per site, but do not reflect actual traffic usage. YouTube is more popular than "www.example.com" yet both will appear as equal value when comparing utilization.

With this in mind, here are a few statistics that were intentionally not measured in the context of a CDN:

1. Time To First Byte (TTFB)
2. CDN Round Trip Time
3. Core Web Vitals
4. Cache Hit versus Cache Miss performance etc.

> While some of these could be measured with our crawl, and others by using the CrUX dataset, the limitations of our crawl and the use of multiple CDNs by some sites, will be difficult to measure and so could be incorrectly reported against the CDN used in out crawl. For these reasons, we have decided not to measure these statistics in this chapter.

Reviewer: I added this sentence to explain why (I think) you choose not to measure these. Please review and let me know if you want to change.

Reviewer 2: Yes, the addition makes sense. Adds more color to the statement.

Reviewer: Just spotted a typo and repeated use of the phrase “our crawl” so let’s use below tweak instead:

```suggestion
While some of these could be measured with HTTP Archive dataset, and others by using the CrUX dataset, the limitations of our methodology and the use of multiple CDNs by some sites, will be difficult to measure and so could be incorrectly attributed. For these reasons, we have decided not to measure these statistics in this chapter.
```

**CDN detection:** This is primarily done through DNS resolution and HTTP headers. Most CDNs use a DNS CNAME to map a user to an optimal datacenter. However, some CDNs use AnyCast IPs or direct A+AAAA responses from a delegated domain which hide the DNS chain. In other cases, websites use multiple CDNs to balance between vendors, which is hidden from the single-request pass of our crawler.

> {# TODO: Should this be in bullet points? I feel like it would be organized a bit better, but open to either way #}

Author: @Navaneeth-akam

Reviewer 2: I would prefer it to be in paragraph style :)

But why do we see this negative growth in CDN Adoption for third party domains?

The probable reasons for this may include:

* The HTTP/2 protocol requires web owners to consolidate the domains instead of using multiple domains for optimal performance
* Contribution of third party content to total page weight has also increased over the years (refer to the [Third Parties](./third-parties) chapter for more details) leading to increased page load time concerns for web owners
* Customization/personalization of third party scripts to suit the requirements of web owners

These changes have led to the SaaS vendors offering "self-hosting" options to web owners. This leads to more content being delivered over the first party domain instead of the vendor's domain. When this happens, it's up to the web owner to either deliver the content over a CDN or directly from their hosting infrastructure.

While we observed CDN adoption across different types of content, we will not look at this data a bit different below.

> {# TODO: I don't get what - "we will not look at this data a bit different below" means exactly #}

Author: @Navaneeth-akam

Reviewer 2: Ahhh....a bit of typo here. It should have been "we will look at this data a bit different below"

  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vSzfIy4p9ujiLHb9T6GpyDE9j7Pni5vBirHhsD53y5my_U4grve1zE4jTWdqGmmXtZahnBOzFeoil52/pubchart?oid=1707360768&format=interactive",
  sheets_gid="414030881",
  sql_file="cdn_usage_by_site_rank.sql"
  )
}}

Ranking the websites based on their popularity (sourced from Google's [Chrome UX Report](./methodology#chrome-ux-report)) in the web and then checking for their CDN usage, the top 1,000 contribute to the highest usage of CDN. The top websites are owned by companies like Google and Amazon, who contribute to much of the internet traffic we see today, so it's no surprise that these names make it to the list of [top CDN providers](#top-CDN-providers) in the next section. This also backs the fact about the benefits CDNs bring to the table when operating at scale and having the ability to scale further if needed.

{{ figure_markup(

>   caption="Percent of top 1000 websites using a CDN.",

Reviewer 3: ```suggestion
  caption="Percent of top 1,000 websites using a CDN.",
```

  )
}}

> Ranking the domains based on their popularity (at the time of this writing sourced from Google CRUX data) in the web and then checking for their CDN usage, the top 1000 websites contribute to the highest usage of CDN. The top websites include domains owned by names like Google and Amazon who contribute to much of the internet traffic we see today, so it's no surprise that these names make it to the list of "Top CDN Providers" in the [next section](#top-CDN-providers). This also backs the fact about the benefits CDNs bring to the table when operating at scale and having the ability to scale further if needed.

Reviewer 3: ```suggestion
Ranking the websites based on their popularity (sourced from Google's [Chrome UX Report](./methodology#chrome-ux-report)) in the web and then checking for their CDN usage, the top 1,000 contribute to the highest usage of CDN. The top websites are owned by companies like Google and Amazon, who contribute to much of the internet traffic we see today, so it's no surprise that these names make it to the list of [top CDN providers](#top-CDN-providers) in the next section. This also backs the fact about the benefits CDNs bring to the table when operating at scale and having the ability to scale further if needed.
```

{{ figure_markup(

>   caption="Sites using CDN among top 10 million sites",

Reviewer 3: @Navaneeth-akam should we also specify that this is usage of a CDN for **hosting** the site (serving the HTML document) to distinguish from CDN usage of other 1P and 3P resources?

```suggestion
  caption="Percent of websites using a CDN.",
```

  content="27%",
  classes="big-number",
  sheets_gid="414030881",

>   sql_file="CDN Usage by Site Popularity.sql"

Reviewer 3: ```suggestion
  sql_file="cdn_usage_by_site_rank.sql"
```

> {{ figure_markup(
>   caption="Sites using CDN among top 10 million sites",
>   content="27%",
>   classes="big-number",
>   sheets_gid="414030881",
>   sql_file="CDN Usage by Site Popularity.sql"
> )
> }}

Reviewer: @Navaneeth-akam we avoid using 10 million number (since it's not). Also SQL file looked incorrect.

```suggestion
{{ figure_markup(
  caption="Sites using CDN.",
  content="27%",
  classes="big-number",
  sheets_gid="414030881",
  sql_file="cdn_usage_by_site_rank.sql"
)
}}
```

Reviewer 2: Understood. I did a change to the big number to show the CDN usage for top 1000 instead of the 10 million number

> The CDN adoption rate falls below 50% when we look at the top 100,000 websites but the rate of reduction slows down beyond this. For the full data set (which is 6.2 million sites on desktop and 7.5 million on mobile), 27% of these websites use CDN. When you translate that percentage into real number, that's 2.7 million websites using CDN! That is not a small number when you look at it this way.

Reviewer 3: ```suggestion
The CDN adoption rate falls below 50% when we look at the top 100,000 websites but the rate of reduction slows down beyond this. For the full data set (which is 6.2 million sites on desktop and 7.5 million on mobile), 27% of these websites use CDN. When you translate that percentage into real number, that's 2.0 million mobile websites using CDN! It's not such a small number when you look at it this way.
```

  )
}}

Ranking the domains based on their popularity (at the time of this writing sourced from Google CRUX data) in the web and then checking for their CDN usage, the top 1000 websites contribute to the highest usage of CDN. The top websites include domains owned by names like Google and Amazon who contribute to much of the internet traffic we see today, so it's no surprise that these names make it to the list of "Top CDN Providers" in the [next section](#top-CDN-providers). This also backs the fact about the benefits CDNs bring to the table when operating at scale and having the ability to scale further if needed.

The CDN adoption rate falls below 50% when we look at the top 100,000 websites but the rate of reduction slows down beyond this. For the full data set (which is 6.2 million sites on desktop and 7.5 million on mobile), 27% of these websites use CDN. When you translate that percentage into real number, that's

2.7 million websites using CDN! That is not a small number when you look at it this way.

> {# TODO: this sentence being broken off seems like a typo. I can merge it back to the previous paragraph but this might also be a good spot to put a big number figure if you want to highlight that number #}

Author: @Navaneeth-akam

Reviewer 2: yes. its a typo. But your suggestion is good. We can put 2.7 million as a big number and highlight it. That will get some attention

Author: I haven't added images before, are you able to commit an update here to add it in?

Reviewer 2: I don't have a big number figure as such. I was thinking more along the lines of increased font size, making them bold and centering it between the paragraph. Does this make sense?

Reviewer 2: Else, we can just make the 2.7 million in bold as a highlight if a figure is not doable

Author: there's a reference for it here https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Figures-Guide#call-out-figures
maybe @tunetheweb can help?

Reviewer 3: @Navaneeth-akam if it exists in the [results sheet](https://docs.google.com/spreadsheets/d/1DL7Pn1vbBwYmQZ5JPAjD69oCOvUidbuoNvdrw--j00U/edit) it's easy to make it into a big number figure.

In the cdn_usage_by_site_rank tab I see 2,012,646 mobile "requests" (should this be "pages"?) that use a CDN in the ALL ranking. Would 2M be more accurate or did you get 2.7M from a different query?

Reviewer 2: i calculated 2.7 million looking at the numbers in the results. You will see that for "Top 10 million site", 27% use CDN. Which translates to 2.7 million sites out of 10 million use CDN.

Reviewer 3: Oh ok there are only 7,491,736 pages in the "all" rank, which is why we avoid calling it 10M.

Reviewer 2: Ahh....ok. Gotcha. Then may be we can take out that big number thing. Because we are now quoting top 10 million sites, so trying to take 27% of 7,491,736 will not be equal to 27% of 10 million, Thus it might be confusing for readers how we came up with this number. Thoughts?

Reviewer: Why not just put 27% in a big number?

Reviewer 2: That thought did cross my mind after i commented :) We can do that. Works for me. 👍🏼

* Google,
* AWS

> {# TODO: these lists feel disruptive when reading. Maybe they can be combined into a sentence or 2? #}

Author: @Navaneeth-akam

Reviewer 2: Yes. This one had mixed opinion from the reviewers. They felt having it as bullet points is better. But i do feel putting them in a sentence makes more sense. We can go ahead and put them in a sentence

For example, a CDN built specifically to deliver advertisements will be optimized for:

- High input-output (I/O) operations
- Effective management of [long tail](https://en.wikipedia.org/wiki/Long_tail) content
- Geographical closeness to businesses requiring their services

> {# TODO aren't those the benefits of all CDNs? #}

Reviewer: @Navaneeth-akam does this need clarification? I was a bit confused.

Reviewer 2: Long tail content management is not something generic CDN's are good at by default. A purpose built CDN can be designed to manage this a bit better. As for geographical closeness, generic CDN will be deployed based on certain assessment but their geographical closeness is relative to Origin. Purpose built CDN will have a focused end user set where they will need deployment of server, which differs a bit from generic CDN

Reviewer: The long tails makes sense (missed that to be honest), and guess that similarly applies to high input/output. I guess for geographical closeness I considered all CDNs to have good global presence but that’s maybe not the case?

Anyway let’s remove the TODO if happy with the wording.

```suggestion
```

This is quite important when you consider that all sites will have to be on TLS these days. Optimal performance at this layer is essential for other steps that follow TLS connection. In this regard, CDNs are able to move more users to lower percentile brackets compared to direct origin connections.

## HTTP/2+ (HTTP/2 or better) adoption

HTTP/2 was introduced with a lot of hype and expectation. This was because the application layer protocol had not been updated since HTTP 1.1 in 1997. Since then, the web traffic trend, content-type, content size, website design, platforms, mobile apps and more have evolved significantly. Thus, there was a need to have a protocol which can meet the requirements of the modern-day web traffic and that protocol was realized with HTTP/2, and then further improved with the more recent HTTP/3.

However, the implementation challenges of HTTP/2 discouraged adoption. In addition, the net performance gains which can be expected with these changes was also not clear. Challenges repeated with the introduction of HTTP/3.

> This was where the CDNs being the intermediary can help in bridging the challenge of HTTP/2 implementation for web owners. An HTTP/2 connection terminates at the CDN level, and this provides web owners the ability to deliver their website and subdomains over HTTP/2 without the need to upgrade their infrastructure to support it—the exact same reasons and benefits we saw for newer TLS versions.

Reviewer: Tweaked this slightly to introduce HTTP/3. Let me know if you want to change.

Reviewer 2: Works for me Barry. I'm good with these tweaks

However, the implementation challenges of HTTP/2 features discouraged adoption. In addition, the net performance gains which can be expected with these changes was also not clear.

This was where the CDNs became an intermediary in bridging the challenge of HTTP/2 implementation for web owners. Similar to TLS, an HTTP/2 connection terminates at the CDN level and this provides web owners the ability to deliver their website and subdomains over HTTP/2 without the need to upgrade their infrastructure to support it.

CDNs acted as the proxy to bridge the gap by providing a layer to consolidate hostnames and route traffic to relevant endpoints with minimal change to their hosting infrastructure. Features like prioritizing content in the queue, server push could be managed from the CDN's side and a few CDN's even provided hands-off automated solutions to run these features without any inputs from website owners, thus providing a push to HTTP/2 adoption (no pun intended ;)).

> {# TODO: are emojis against our editors guide? leaving a note here to follow up #}

Author: @tunetheweb thoughts on the line above where it says (no pun intended ;))?

Reviewer 2: Not sure about the policy for this. I'm other either way

Reviewer: I say remove if you don't mind.

  )
}}

> On both desktop and Mmbile platforms, we see that CDNs are delivering twice as many text assets in Brotli, unlike domains delivered from origin.  From the [CDN Adoption](#cdn-adoption) section covered earlier, 73% of the domains serving HTMLs are on CDNs and these all  benefit from the Brotli compression. By offloading the computational load of compressing a text asset in the Brotli format to CDNs, website owners need not invest resources for hosting infrastructure.

Reviewer 3: ```suggestion
On both desktop and mobile platforms, we see that CDNs are delivering twice as many text assets in Brotli, unlike domains delivered from origin.  From the [CDN Adoption](#cdn-adoption) section covered earlier, 73% of the domains serving HTMLs are on CDNs and these all  benefit from the Brotli compression. By offloading the computational load of compressing a text asset in the Brotli format to CDNs, website owners need not invest resources for hosting infrastructure.
```
